File: python/xviz_avs/builder/time_series.py
from xviz_avs.builder.base_builder import XVIZBaseBuilder, CATEGORY
from xviz_avs.v2.core_pb2 import TimeSeriesState

class XVIZTimeSeriesBuilder(XVIZBaseBuilder):
    def __init__(self, metadata, logger=None):
        super().__init__(CATEGORY.TIME_SERIES, metadata, logger)

        # Stores time_series data by timestamp then id
        # They will then be group when constructing final object
        self._data = {}
        self._reset()

    def id(self, identifier):
        self._validate_prop_set_once('_id')
        self._id = identifier
        return self

    def value(self, value):
        self._validate_prop_set_once('_value')

        if isinstance(value, list):
            self._logger.error("Input `value` must be single value")

        self._value = value
        return self

    def timestamp(self, timestamp):
        self._validate_prop_set_once('_timestamp')

        if isinstance(timestamp, list):
            self._logger.error("Input `value` must be single value")

        self._timestamp = timestamp
        return self

    def get_data(self):
        self._flush()
        if not self._data:
            return None

        time_series_data = []
        for timestamp, ids in self._data.items():
            for id_, fields in ids.items():
                for tsdata in fields.values():
                    entry = TimeSeriesState(
                        timestamp=timestamp,
                        streams=tsdata['streams'],
                        values=tsdata['values'],
                        object_id=id_
                    )

                    time_series_data.append(entry)

        return time_series_data

    def _add_timestamp_entry(self):
        # this._data structure
        # timestamp: {
        #   id: {
        #     fieldName: {
        #       streams: []
        #       values: []
        #     }
        #   }
        # }
        if not self._data_pending():
            return

        if isinstance(self._value, str):
            field_name = "strings"
        elif isinstance(self._value, bool):
            field_name = "bools"
        elif isinstance(self._value, int):
            field_name = "int32s"
        elif isinstance(self._value, float):
            field_name = "doubles"
        else:
            self._logger.error("The type of input value is not recognized!")
        
        ts_entry = self._data.get(self._timestamp)
        if ts_entry:
            id_entry = ts_entry.get(self._id)
            if id_entry:
                field_entry = id_entry.get(field_name)
                if field_entry:
                    field_entry['streams'].append(self._stream_id)
                    field_entry['values'][field_name].append(self._value)
                else:
                    id_entry[field_name] = self._get_field_entry(field_name)
            else:
                ts_entry[self._id] = self._get_id_entry(field_name)
        else:
            ts_entry = {self._id: self._get_id_entry(field_name)}
            self._data[self._timestamp] = ts_entry

    def _get_id_entry(self, field_name):
        return {field_name: self._get_field_entry(field_name)}

    def _get_field_entry(self, field_name):
        return dict(streams=[self._stream_id], values={field_name: [self._value]})

    def _data_pending(self):
        return self._value or self._timestamp or self._id

    def _validate(self):
        if self._data_pending():
            super()._validate()

            self._validate_has_prop("_value")
            self._validate_has_prop("_timestamp")

    def _flush(self):
        self._validate()
        self._add_timestamp_entry()
        self._reset()

    def _reset(self):
        self._id = None
        self._value = None
        self._timestamp = None

File: cpp/src/builder/time_series.cpp
/*
 * File: time_series.cpp
 * Author: Minjun Xu (mjxu96@gmail.com)
 * File Created: Saturday, 4th January 2020 5:18:49 am
 */

#include "xviz/builder/time_series.h"

using namespace xviz;

using TimeSeriesFieldEntryType = std::pair<std::vector<std::string>, std::unordered_map<std::string, std::vector<std::variant<std::string, bool, int, double>>>>;
using TimeSeriesFieldType = std::unordered_map<std::string, TimeSeriesFieldEntryType>;
using TimeSeriesIdType = std::unordered_map<std::string, TimeSeriesFieldType>;
using TimeSeriesType = std::unordered_map<double, TimeSeriesIdType>;

XVIZTimeSeriesBuilder::XVIZTimeSeriesBuilder(const std::shared_ptr<Metadata>& metadata) : 
  XVIZBaseBuilder(Category::StreamMetadata_Category_TIME_SERIES, metadata) {
  data_ = std::make_shared<TimeSeriesType>();
  Reset();
}

void XVIZTimeSeriesBuilder::DeepCopyFrom(const XVIZTimeSeriesBuilder& other) {
  XVIZBaseBuilder::DeepCopyFrom(other);
  DeepCopyPtr(data_, other.data_);
  DeepCopyPtr(id_, other.id_);
  DeepCopyPtr(value_, other.value_);
  DeepCopyPtr(timestamp_, other.timestamp_);
  vpos_ = other.vpos_;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Stream(const std::string& stream_id) {
  if (stream_id_.size() > 0) {
    Flush();
  }

  stream_id_ = stream_id;
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Id(const std::string& id) {
  id_ = std::make_shared<std::string>(id);
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Id(std::string&& id) {
  id_ = std::make_shared<std::string>(std::move(id));
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Value(const std::string& value) {
  value_ = std::make_shared<std::variant<std::string, bool, int, double>>(value);
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Value(std::string&& value) {
  value_ = std::make_shared<std::variant<std::string, bool, int, double>>(std::move(value));
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Value(const char* value) {
  value_ = std::make_shared<std::variant<std::string, bool, int, double>>(std::string(value));
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Value(bool value) {
  value_ = std::make_shared<std::variant<std::string, bool, int, double>>(value);
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Value(int value) {
  value_ = std::make_shared<std::variant<std::string, bool, int, double>>(value);
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Value(double value) {
  value_ = std::make_shared<std::variant<std::string, bool, int, double>>(value);
  return *this;
}

XVIZTimeSeriesBuilder& XVIZTimeSeriesBuilder::Timestamp(double timestamp) {
  timestamp_ = std::make_shared<double>(timestamp);
  return *this;
}

std::shared_ptr<std::vector<TimeSeriesState>> XVIZTimeSeriesBuilder::GetData() {
  FlushNotReset();

  auto ts_state = std::make_shared<std::vector<TimeSeriesState>>();

  // if (IsDataPending()) {
  //   AddTimestampEntry();
  //   Reset();
  //   return ts_state;
  // }

  for (const auto& [timestamp, ids] : *data_) {
    for (const auto& [id, fields] : ids) {
      for (const auto& [field_name, entry] : fields) {
        TimeSeriesState ts;
        ts.set_object_id(id);
        ts.set_timestamp(timestamp);
        for (const auto& stream : std::get<0>(entry)) {
          auto new_stream_ptr = ts.add_streams();
          // TODO is it correct?
          *new_stream_ptr = stream;
        }

        auto value_ptr = ts.mutable_values();
        for (auto& [fn, values] : std::get<1>(entry)) {
          for (auto& value : values) {
            auto vpos = value.index();
            switch (vpos) {
              case 0u:
                value_ptr->add_strings(std::get<0u>(value));
                break;
              case 1u:
                value_ptr->add_bools(std::get<1u>(value));
                break;
              case 2u:
                value_ptr->add_int32s(std::get<2u>(value));
                break;
              case 3u:
                value_ptr->add_doubles(std::get<3u>(value));
                break;
              default:
                XVIZ_LOG_ERROR("The type of input value is not reconized");
            }
          }
        }
        ts_state->emplace_back(std::move(ts));
      }
    }
  }

  Reset();

  return ts_state;
}

void XVIZTimeSeriesBuilder::Validate() {
  XVIZBaseBuilder::Validate();
}

void XVIZTimeSeriesBuilder::AddTimestampEntry() {
  // if (!IsDataPending()) {
  //   return;
  // }
  std::string field_name;

  vpos_ = value_->index();

  switch (vpos_) {
    case 0u:
      field_name = "strings";
      break;
    case 1u:
      field_name = "bools";
      break;
    case 2u:
      field_name = "int32s";
      break;
    case 3u:
      field_name = "doubles";
      break;
    default:
      XVIZ_LOG_ERROR("The type of input value is not reconized");
      return;
  }

  if (data_->find(*timestamp_) != data_->end()) {
    TimeSeriesIdType& ts_entry = (*data_)[*timestamp_];
    if (ts_entry.find(*id_) != ts_entry.end()) {
      TimeSeriesFieldType& id_entry = ts_entry[*id_];
      if (id_entry.find(field_name) != id_entry.end()) {
        std::get<0>(id_entry[field_name]).emplace_back(stream_id_);
        std::get<1>(id_entry[field_name])[field_name].emplace_back(*value_);
      } else {
        id_entry[field_name] = GetFieldEntry(field_name);
      }
    } else {
      ts_entry[*id_] = GetIdEntry(field_name);
    }
  } else {
    (*data_)[*timestamp_] = TimeSeriesIdType{
      {*id_, GetIdEntry(field_name)}
    };
  }
}

void XVIZTimeSeriesBuilder::Flush() {
  if (!IsDataPending()) {
    return;
  }
  Validate();
  AddTimestampEntry();
  Reset();
}

void XVIZTimeSeriesBuilder::FlushNotReset() {
  if (!IsDataPending()) {
    return;
  }
  Validate();
  AddTimestampEntry();
}

void XVIZTimeSeriesBuilder::Reset() {
  id_ = nullptr;
  value_ = nullptr;
  timestamp_ = nullptr;
}

bool XVIZTimeSeriesBuilder::IsDataPending() {
  return id_ != nullptr && timestamp_ != nullptr && value_ != nullptr;
}

TimeSeriesFieldEntryType XVIZTimeSeriesBuilder::GetFieldEntry(const std::string& field_name) {
  TimeSeriesFieldEntryType entry;
  std::get<0>(entry).emplace_back(stream_id_);
  std::get<1>(entry)[field_name].emplace_back(*value_);
  return entry;
}

TimeSeriesFieldType XVIZTimeSeriesBuilder::GetIdEntry(const std::string& field_name) {
  return TimeSeriesFieldType{
    {field_name, GetFieldEntry(field_name)}
  };
}